File: CSharp Multi-Agent/Multi-Agent Lab/MultiAgentLibrary/Field.cs
//***********************************************************************
// Assembly         : MultiAgentLibrary
// Author           : Tony Richards
// Created          : 08-15-2011
//
// Last Modified By : Tony Richards
// Last Modified On : 08-18-2011
// Description      : 
//
// Copyright (c) 2011, Tony Richards
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
//
// Redistributions of source code must retain the above copyright notice, this list
// of conditions and the following disclaimer.
//
// Redistributions in binary form must reproduce the above copyright notice, this
// list of conditions and the following disclaimer in the documentation and/or other
// materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
// OF THE POSSIBILITY OF SUCH DAMAGE.
//***********************************************************************
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;

namespace MultiAgentLibrary
{
    using System.Drawing;
    using System.Threading.Tasks;
    using System.Globalization;
    using System.Xml.Serialization;

    [XmlRoot]
    public class Field
    {
        [XmlElement]
        public Point StartPoint { get; set; }

        [XmlIgnore]
        public List<Agent> AgentsList { get; set; }

        [XmlArray]
        [XmlArrayItem("P")]
        public List<Point> OriginalRoute { get; set; }

        [XmlIgnore]
        public List<Point> ShortestRoute { get; set; }

        [XmlArray]
        [XmlArrayItem("FS")]
        public FieldSquare[] Squares { get; set; }

        [XmlAttribute]
        public int Width { get; set; }

        [XmlAttribute]
        public int Height { get; set; }

        [XmlElement]
        public SizeF SquareSize { get; set; }

        public Field()
        {
            AgentsList = new List<Agent>();
            OriginalRoute = new List<Point>();
            ShortestRoute = new List<Point>();
        }

        public Field(int width) : this(width, width) { }

        public Field(int width, int height)
            : this()
        {
            Squares = new FieldSquare[width * height];

            for (var x = 0; x < width; x++)
            {
                for (var y = 0; y < height; y++)
                {
                    Squares[x + y * width] = new FieldSquare(new Point(x, y));
                }
            }
            Width = width;
            Height = height;
        }

        public Field(int width, int height, string fileName)
            : this(width, height)
        {
            if (width == 0)
                throw new ArgumentOutOfRangeException("width");
            if (height == 0)
                throw new ArgumentOutOfRangeException("height");

            // File Format : 
            //
            // x,y,sensor0ang,sensor0val,sensor0state,sensor1ang,sensor1val,sensor1state,...,sensor39ang,sensor39val,sensor39state
            // x,y,sensor0ang,sensor0val,sensor0state,sensor1ang,sensor1val,sensor1state,...,sensor39ang,sensor39val,sensor39state
            //
            // Repeat as required...
            //
            // x, y and val values are floats.
            // state values should be strings.
            //
            // sensornval is a floating point number.

            // Read the file in to memory.
            // Convert to a list to make sure all the reading is done now, not on demand.
            var lines = File.ReadAllLines(fileName);

            // Work out the starting point.
            var firstLineParts = lines.First().Split(new[] { "," }, 3, StringSplitOptions.RemoveEmptyEntries);
            var origStartPoint = new PointF(float.Parse(firstLineParts[0], CultureInfo.InvariantCulture), float.Parse(firstLineParts[1], CultureInfo.InvariantCulture));

            var unscaledOriginalRoute = new List<PointF>();

            // Split each line up in to a list of readings, then merge them all together.
            // Don't care about order, so can be done in parallel.

            var rawPoints = lines.AsParallel().SelectMany(l =>
                {
                    var parts = l.Split(new[] { "," }, StringSplitOptions.RemoveEmptyEntries);
                    var origin = new PointF(float.Parse(parts[0], CultureInfo.InvariantCulture), float.Parse(parts[1], CultureInfo.InvariantCulture));

                    unscaledOriginalRoute.Add(origin);

                    var output = new List<SensorReading>();

                    for (var i = 2; i < parts.Length; i += 3)
                    {
                        var reading = new SensorReading
                        {
                            Angle = float.Parse(parts[i], CultureInfo.InvariantCulture),
                            Range = float.Parse(parts[i + 1], CultureInfo.InvariantCulture),
                            State = (SensorState)Enum.Parse(typeof(SensorState), parts[i + 2], true),
                            Origin = origin
                        };

                        output.Add(reading);
                    }
                    return output;
                }).Select(r => new
                                    {
                                        Point = new PointF(r.Origin.X + r.Range * (float)Math.Sin(r.Angle), r.Origin.Y + r.Range * (float)Math.Cos(r.Angle)),
                                        r.State
                                    });

            // Work out the dimensions of the map
            var maxX = rawPoints.Max(p => p.Point.X);
            var maxY = rawPoints.Max(p => p.Point.Y);

            // Get the size of each new map square
            var yRectSize = maxY / (height - 1);
            var xRectSize = maxX / (width - 1);
            SquareSize = new SizeF(xRectSize, yRectSize);

            // Scale the starting point
            StartPoint = new Point((int)Math.Round(origStartPoint.X / xRectSize), (int)Math.Round(origStartPoint.Y / yRectSize));

            OriginalRoute.AddRange(unscaledOriginalRoute.Select(p => new Point((int)Math.Round(p.X / xRectSize), (int)Math.Round(p.Y / yRectSize))));

            Parallel.ForEach(rawPoints, p =>
                {
                    var newPoint = new Point((int)Math.Round(p.Point.X / xRectSize), (int)Math.Round(p.Point.Y / yRectSize));

                    var index = newPoint.X + newPoint.Y * height;

                    lock (Squares[index].LockObject)
                    {
                        if (p.State == SensorState.End)
                        {
                            Squares[index].SquareType = SquareType.Destination;
                        }
                        else if (p.State == SensorState.Boundary && Squares[index].SquareType != SquareType.Destination)
                            Squares[index].SquareType = SquareType.Wall;
                    }
                });
        }

        public void CycleAgents()
        {
            Parallel.ForEach(AgentsList, agent => agent.Process(this));

            Parallel.ForEach(Squares.Where(square => square.PheromoneLevel > 1 && square.SquareType == SquareType.Passable),
                square =>
                {
                    square.PheromoneLevel -= FieldSquare.PheromoneDecayRate;
                });
        }
    }
}

File: trunk/MultiAgent/Classes/Field.cpp
/* 
 * File:   Field.cpp
 * Author: Tony
 * 
 * Created on 03 November 2010, 22:04
 */

#include <omp.h>

#include "Field.h"
#include "StringTools.h"
#include "Point.h"
#include <fstream>
#include <vector>
#include <stdlib.h>
#include <math.h>

using namespace std;

enum SensorState
{
    Boundary,
    End,
    None
};

struct SensorReading
{
public:
    PointF Point;
    SensorState State;
};

Field::Field(const int width, const int height, const string filename)
{
    int size = width * height;
    squares = new FieldSquare[size];

    vector<string> lines;
    ifstream dataFile(filename.c_str());
    if (dataFile.is_open())
    {
        string line;
        while (dataFile.good())
        {
            getline (dataFile, line);
            lines.push_back(line);
        }
        if (lines.size() == 0)
        {
            InvalidFileException exp;
            throw exp;
        }

        dataFile.close();
    }

    vector<SensorReading> readings;

    float maxX = 0.0;
    float maxY = 0.0;
    PointF startPointF;

    for (int i = 0; i < lines.size(); i++)
    {
        string line = lines[i];
        vector<string> parts;
        StringTools::Tokenize(line, parts);
        PointF origin(atof(parts[0].c_str()), atof(parts[1].c_str()));

        if (i == 0)
            startPointF = origin;

        for (int j = 2; j < parts.size(); j += 3)
        {
            SensorState state;

            if (parts[i+2] == "boundary")
                state = Boundary;
            else if (parts[i+2] == "end")
                state = End;
            else
                state = None;

            if (state == None)
            {
                // A useless reading that can be ignored.
                continue;
            }

            double angle(atof(parts[i].c_str()));
            double range(atof(parts[i+1].c_str()));

            PointF point(origin.X + range * sin(angle), origin.Y + range * cos(angle));
            
            SensorReading reading;
            reading.Point = point;
            reading.State = state;

            if (point.X > maxX)
                maxX = point.X;
            if (point.Y > maxY)
                maxY = point.Y;

            readings.push_back(reading);
        }
    }

    float xRectSize = maxX / width;
    float yRectSize = maxY / height;

    int startPointX = (int)round(startPointF.X / xRectSize);
    int startPointY = (int)round(startPointF.Y / yRectSize);

    this->startPoint = Point(startPointX, startPointY);

    for (int i = 0; i < readings.size(); i++)
    {
        int scalePointX = (int)round(readings[i].Point.X / xRectSize);
        int scalePointY = (int)round(readings[i].Point.Y / yRectSize);

        int index = scalePointX + scalePointY * width;

        if (squares[index].GetPosition() == Point(scalePointX,scalePointY))
        {
            squares[index].SetPosition(Point(scalePointX, scalePointY));
            switch (readings[i].State)
            {
                case Boundary:
                    squares[index].SetType(Wall);
                    break;
                case End:
                    squares[index].SetType(Destination);
                    break;
                default:
                    squares[index].SetType(Passable);
                    break;
            }
        }
        else
        {
            switch (readings[i].State)
            {
                case Boundary:
                    if (squares[index].GetType() == Passable)
                        squares[index].SetType(Wall);
                    break;
                case End:
                    squares[index].SetType(Destination);
                    break;
            }
        }
    }
}

Field::~Field()
{
    delete[] agents;
    agents = NULL;
}

